compiler-cli/src/dependencies.rs

```diff
@@ -492,38 +516,230 @@ fn resolve_versions<Telem: Telemetry>(
+        let version = match recipe {
+            Recipe::Hex { version } => version,
+            Recipe::Path { path } => {
+                provide_local_package(&name, &path, &mut provider_info, &mut provided_packages)?
+            }
+            Recipe::Git { git } => {
+                provide_git_package(&name, &git, &mut provider_info, &mut provided_packages)?
+            }
+        };
+        let _ = version_requirements.insert(name, version);
+    }
+
+    let resolved = dependency::resolve_versions(
         PackageFetcher::boxed(runtime.clone()),
-        mode,
-        config,
-        manifest,
+        provided_packages.clone(),
+        config.name.to_string(),
+        version_requirements.into_iter(),
+        &locked,
     )?;
-    let packages = runtime.block_on(future::try_join_all(
-        resolved
-            .into_iter()
-            .map(|(name, version)| lookup_package(name, version)),
-    ))?;
+
+    let provided_package_requirements = provided_packages
+        .into_iter()
+        .map(|(name, package)| {
+            (
+                name,
+                package.releases[0].requirements.keys().cloned().collect(),
+            )
+        })
+        .collect();
+
+    let packages = runtime.block_on(future::try_join_all(resolved.into_iter().map(
+        |(name, version)| {
+            lookup_package(
+                name,
+                version,
+                &provider_info,
+                &provided_package_requirements,
+            )
+        },
+    )))?;
     let manifest = Manifest {
         packages,
         requirements: config.all_dependencies()?,
     };
     Ok(manifest)
 }
 
-async fn lookup_package(name: String, version: Version) -> Result<ManifestPackage> {
-    let config = hexpm::Config::new();
-    let release = hex::get_package_release(&name, &version, &config, &HttpClient::new()).await?;
-    let manifest = ManifestPackage {
-        name,
-        version,
-        otp_app: Some(release.meta.app),
-        build_tools: release.meta.build_tools,
-        requirements: release.requirements.keys().cloned().collect_vec(),
-        source: ManifestPackageSource::Hex {
-            outer_checksum: Base16Checksum(release.outer_checksum),
-        },
+#[derive(Clone, Eq, PartialEq, Debug)]
+enum ProviderInfo {
+    Git { repo: String, commit: String },
+    Local { path: PathBuf },
+}
+
+fn provide_local_package(
+    package_name: &str,
+    package_path: &Path,
+    info: &mut HashMap<String, ProviderInfo>,
+    provided: &mut HashMap<String, hexpm::Package>,
+) -> Result<hexpm::version::Range> {
+    let canonical_path = package_path
+        .canonicalize()
+        .expect("local package cannonical path");
+    let package_info = ProviderInfo::Local {
+        path: canonical_path.clone(),
     };
-    Ok(manifest)
+
+    // Determine if package has already been walked
+    match info.insert(package_name.to_string(), package_info.clone()) {
```

Comment:
Let's use SmolStr here.

Reply:
This would require changing `package_name` to `&SmolStr` but `package_name` is called with dependency names from the config where the type is `Dependencies = HashMap<String, Recipe>`. I can change this definition but I get worrying errors from in codegen. Otherwise I could call `SmallStr::from<String>` but it seems kind of contrived.